SilKit/source/services/orchestration/TimeSyncService.cpp

```diff
@@ -546,6 +612,95 @@ bool TimeSyncService::AbortHopOnForCoordinatedParticipants() const
     }
     return false;
 }
+
+auto TimeSyncService::GetCurrentWallClockSyncPoint() const -> std::chrono::nanoseconds
+{
+    Lock lock{_mx};
+    return _currentWallClockSyncPoint;
+}
+
+
+// Mixture of sleep_for and busy waiting to achieve higher precision sleeps with the low-res windows times
+void TimeSyncService::HybridWait(std::chrono::nanoseconds targetWaitDuration)
+{
+    auto busyWait = [](std::chrono::nanoseconds duration) {
+        auto startTime = std::chrono::steady_clock::now();
+        while (std::chrono::steady_clock::now() - startTime < duration)
+        {
+            std::this_thread::yield();
+        }
+    };
+
+    // By default, Windows timer have a resolution of 15.6ms
+    // The effective time that wait_for or sleep_for actually waits will be a step function with steps every 15.6ms
+    const auto defaultTimerResolution = 16ms; 
+
+    if (targetWaitDuration < defaultTimerResolution)
+    {
+        busyWait(targetWaitDuration);
+    }
+    else
+    {
+        auto timeBeforeSleep = std::chrono::steady_clock::now();
+        // Wait the share of the targetWaitDuration that we can precisely achieve with the low timer resolution
+        auto sleepDuration = targetWaitDuration - defaultTimerResolution; 
+        std::this_thread::sleep_for(sleepDuration);
+        // Busy-wait the remainder
+        auto remainder = targetWaitDuration - (std::chrono::steady_clock::now() - timeBeforeSleep);
+        busyWait(remainder);
+    }
+}
+
+
+void TimeSyncService::StartWallClockCouplingThread()
+{
+    _wallClockCouplingThreadRunning = true;
+    _wallClockCouplingThread = std::thread{[this]() {
+
+        SilKit::Util::SetThreadName("SilKit-WallClockCoupling");
+
+        const auto startTime = std::chrono::steady_clock::now();
+        auto nextAnimatedWallClockSyncPoint = _timeConfiguration.NextSimStep().duration * _animationFactor;
+
+        while (_wallClockCouplingThreadRunning)
+        {
+            // Wait until the next SimTask is due according to the (animated) wall clock
+            auto currentRunDuration = std::chrono::steady_clock::now() - startTime;
+            auto durationUntilNextWallClockSyncPoint = nextAnimatedWallClockSyncPoint - currentRunDuration;
+            HybridWait(std::chrono::duration_cast<std::chrono::nanoseconds>(durationUntilNextWallClockSyncPoint));
+
+            {
+                Lock lock{_mx};
```

Comment:
If I'm not mistaken, the mutex `_mx` is only used to protect access to the `_currentWalLClockSyncPoint`. A `std::atomic<std::chrono::nanoseconds::rep> _currentWalLClockSyncPointNs` would do the job too.

Reply:
It's also for the `_timeConfiguration.NextSimStep()` which might change at any time.